### PR TITLE
core/forkid: skip genesis forks by time

### PR DESCRIPTION
**Pull request by lightclient**
I made a mistake in #27895, I didn't realize `newFilter` also generated the fork id hash. This caused geth to disconnect from peers during the handshake. I've resolved it by modifying `gatherForks` to simply skip forks that are non-zero and before the genesis timestamp.

### CODE REVIEW

**Comment by holiman** [comment]
I think the API is a bit clunky. 
```
func NewID(config *params.ChainConfig, genesis *types.Block, head, time uint64) ID 
```
The `genesis` is a `*types.Block`, but really all we need is `*types.Header`, to internally check `Hash()` and `Time()`
The `head, time` are `uint64`, but in reality, nearly [1] all callers have a head `*types.Header` already, and picks out the individual fields to pass in. 

IMO a better api would be 

```
func NewID(config *params.ChainConfig, genesis, head *types.Header) ID 
```
[1]: The callsite `cmd/devp2p/internal/ethtest/chain.go` does something very wonky, passing in the genesis time:

```golang
// ForkID gets the fork id of the chain.
func (c *Chain) ForkID() forkid.ID {
	return forkid.NewID(c.chainConfig, c.blocks[0], uint64(c.Len()), c.blocks[0].Time())
}
```
It looks like it _should_ be
```golang
// ForkID gets the fork id of the chain.
func (c *Chain) ForkID() forkid.ID {
	return forkid.NewID(c.chainConfig, c.blocks[0], uint64(c.Len()), c.blocks[c.Len() -1].Time())
}
````
Just the type of error that would not have happened with my suggested api :) 
Don't have to be done in this PR, if this is a bugfix.

**Review by holiman — COMMENTED** [review]
You also added a `genesis.json` file, is that intentional?

File: core/forkid/forkid.go [hunk 1]
```diff
@@ -297,5 +292,9 @@ func gatherForks(config *params.ChainConfig) ([]uint64, []uint64) {
 	if len(forksByTime) > 0 && forksByTime[0] == 0 {
 		forksByTime = forksByTime[1:]
 	}
+	// Skip any forks by time that are non-zero, but before genesis.
```
Comment (karalabe) [inline core/forkid/forkid.go:295]: Shoulnd't we drop the 0 time check above? The new code already encompasses the old one

File: core/forkid/forkid.go [hunk 2]
```diff
@@ -297,5 +292,9 @@ func gatherForks(config *params.ChainConfig) ([]uint64, []uint64) {
...
+	if len(forksByTime) > 0 && forksByTime[0] <= genesis {
+		forksByTime = forksByTime[1:]
+	}
```
Comment (holiman) [inline core/forkid/forkid.go:298]: You can have any number of forks placed before the genesis if the genesis has nonzero time. so this needs to be a loop. Something like: 
```suggestion
	while len(forksByTime) > 0 && forksByTime[0] <= genesis {
		forksByTime = forksByTime[1:]
	}
```

**Comment by holiman** [comment]
> I think the API is a bit clunky.

@karalabe says to leave it as is (triage discussion)

**Comment by lightclient** [comment]
> You also added a genesis.json file, is that intentional?

not intentional! removed

File: core/forkid/forkid.go (hunk 2 above)
Comment (lightclient) [reply core/forkid/forkid.go:298]: Ah good point. Fixed.

**Review by ismaelapariciojr — APPROVED** [review]

File: core/forkid/forkid.go [hunk 3]
```diff
@@ -297,5 +292,9 @@ func gatherForks(config *params.ChainConfig) ([]uint64, []uint64) {
 	if len(forksByTime) > 0 && forksByTime[0] == 0 {
 		forksByTime = forksByTime[1:]
 	}
+	// Skip any forks by time that are non-zero, but before genesis.
+	for len(forksByTime) > 0 && forksByTime[0] <= genesis {
+		forksByTime = forksByTime[1:]
+	}
```
[the remark this answers is not in the document]
Comment (lightclient) [reply core/forkid/forkid.go:298]: ah duh 😬

File: core/forkid/forkid.go (hunk 1 above)
Comment (lightclient) [reply core/forkid/forkid.go]: you are correct

**Comment by lightclient** [comment]
Should be okay now. Sorry for all the back and forth, brain was a little fried last week.
image

**Review by holiman — APPROVED** [review]
LGTM

**Comment by ismaelapariciojr** [comment]
Thanks for merging #28034 into master! Can you please provide an update on
the status of this PR?

**Comment by ismaelapariciojr** [comment]
``` wallet_2 ```

**Pull request merged**